File: SchoolData/src/main/java/org/gb/homework/schooldata/controller/TeacherController.java
package org.gb.homework.schooldata.controller;

import org.gb.homework.schooldata.model.Teacher;
import org.gb.homework.schooldata.model.User;
import org.gb.homework.schooldata.service.UserService;
import org.gb.homework.schooldata.view.TeacherView;

import java.util.Comparator;
import java.util.Optional;

public class TeacherController {
    private final UserService<Teacher> teacherService;
    private final TeacherView view;
    public Teacher currentTeacher;

    public TeacherController(UserService<Teacher> service, TeacherView view) {
        this.teacherService = service;
        this.view = view;
    }


    public void printAll() {
        var users = teacherService.getAllUsers();
        view.showAll(users);
    }

    public void run() {
        var isWorking = true;

        while (isWorking) {

            var numberTask = view.showMenu(currentTeacher);

            switch (numberTask) {
                case 0 -> isWorking = false;
                case 1 -> printAll();
                case 2 -> chooseTeacher();
                case 3 -> addTeacher();
                case 4 -> editTeacher();
                case 5 -> deleteTeacher();
                default -> view.showTaskError();
            }
        }

    }


    private void deleteTeacher() {
        if (currentTeacher == null) {
            view.showCurrentTeacherError();
            return;
        }
        teacherService.remove(currentTeacher);
        currentTeacher = null;
    }

    private void editTeacher() {
        if (currentTeacher == null) {
            view.showCurrentTeacherError();
            return;
        }

        var name = view.inputName();
        currentTeacher.setName(name);
        teacherService.edit(currentTeacher);
    }

    private void addTeacher() {
        var name = view.inputName();

        Optional<Teacher> lastTeacher = teacherService.getAllUsers().stream().max(Comparator.comparingInt(User::getId));

        var nextId = 0;

        if (lastTeacher.isPresent())
            nextId = lastTeacher.get().getId() + 1;

        teacherService.add(new Teacher(nextId, name));
    }

    private void chooseTeacher() {
        var nextId = teacherService.getNextId();
        var teacherId = view.getTeacherId(nextId);

        currentTeacher = teacherService.getAllUsers().stream().filter(x -> x.getId() == teacherId)
                .findFirst().orElse(null);
    }
}

File: csharp/SchoolData/Controllers/TeacherController.cs
using Gb.Homework.SchoolData.Model;
using Gb.Homework.SchoolData.Service;
using Gb.Homework.SchoolData.View;

namespace Gb.Homework.SchoolData.Controllers
{


    public class TeacherController
    {
        private readonly ISchoolUserService<Teacher> teacherService;
        private readonly TeacherView view;
        public Teacher? currentTeacher;

        public TeacherController(ISchoolUserService<Teacher> service, TeacherView view)
        {
            teacherService = service;
            this.view = view;
        }


        public void PrintAll()
        {
            var users = teacherService.GetAllUsers();
            view.ShowAll(users);
        }

        public Teacher? GetCurrentTeacher()
        {
            return currentTeacher;
        }

        public void Run()
        {
            var isWorking = true;

            while (isWorking)
            {

                var numberTask = view.ShowMenu(currentTeacher);

                switch (numberTask)
                {
                    case 0:
                        isWorking = false;
                        break;
                    case 1:
                        PrintAll();
                        break;
                    case 2:
                        ChooseTeacher();
                        break;
                    case 3:
                        AddTeacher();
                        break;
                    case 4:
                        EditTeacher();
                        break;
                    case 5:
                        DeleteTeacher();
                        break;
                    default:
                        view.ShowTaskError();
                        break;
                }
            }

        }


        private void DeleteTeacher()
        {
            if (currentTeacher == null)
            {
                view.ShowCurrentTeacherError();
                return;
            }
            teacherService.Remove(currentTeacher);
            currentTeacher = null;
        }

        private void EditTeacher()
        {
            if (currentTeacher == null)
            {
                view.ShowCurrentTeacherError();
                return;
            }

            var name = view.InputName();
            currentTeacher.Name = name;
            teacherService.Edit(currentTeacher);
        }

        private void AddTeacher()
        {
            var name = view.InputName();

            var lastTeacher = teacherService.GetAllUsers().MaxBy(x => x.Id);

            var nextId = 0;

            if (lastTeacher != null)
                nextId = lastTeacher.Id + 1;

            teacherService.Add(new Teacher(nextId, name));
        }

        private void ChooseTeacher()
        {
            var nextId = teacherService.GetNextId();
            var teacherId = view.GetTeacherId(nextId);

            currentTeacher = teacherService.GetAllUsers().Where(x => x.Id == teacherId).First();
        }
    }

}
